central/reports/scheduler/v2/schedule.go

```diff
@@ -0,0 +1,646 @@
+		dataStartTime, err = s.lastSuccesfulScheduledReportTime(req.Ctx, req.ReportConfig)
+		if err != nil {
+			s.logErrorAndUpsertReportStatus(errors.Wrap(err, "Error finding last successful scheduled report time"), req)
+			return
+		}
+	} else if req.ReportConfig.GetVulnReportFilters().GetSinceStartDate() != nil {
+		dataStartTime = req.ReportConfig.GetVulnReportFilters().GetSinceStartDate()
+	}
+
+	// Get the results of running the report query
+	reportData, err := s.getReportData(req.Ctx, req.ReportConfig, dataStartTime)
+	if err != nil {
+		s.logErrorAndUpsertReportStatus(err, req)
+		return
+	}
+
+	// Format results into CSV
+	zippedCSVData, err := common.Format(reportData)
+	if err != nil {
+		s.logErrorAndUpsertReportStatus(err, req)
+		return
+	}
+
+	// If it is an empty report, do not send an attachment in the final notification email and the email body
+	// will indicate that no vulns were found
+	templateStr := vulnReportEmailTemplate
+	if zippedCSVData == nil {
+		// If it is an empty report, the email body will indicate that no vulns were found
+		templateStr = noVulnsFoundEmailTemplate
+	}
+
+	messageText, err := formatMessage(dataStartTime, templateStr)
+	if err != nil {
+		s.logErrorAndUpsertReportStatus(errors.Wrap(err, "error formatting the report email text"), req)
+		return
+	}
+
+	errorList := errorhelpers.NewErrorList("Error sending email notifications: ")
+	for _, notifierConfig := range req.ReportConfig.GetNotifiers() {
+		nf := s.notificationProcessor.GetNotifier(req.Ctx, notifierConfig.GetEmailConfig().GetNotifierId())
+		reportNotifier, ok := nf.(notifiers.ReportNotifier)
+		if !ok {
+			errorList.AddError(errors.Errorf("incorrect type of notifier '%s'", notifierConfig.GetEmailConfig().GetNotifierId()))
+			continue
+		}
+		err := s.retryableSendReportResults(req.Ctx, reportNotifier, notifierConfig.GetEmailConfig().GetMailingLists(),
+			zippedCSVData, messageText)
+		if err != nil {
+			errorList.AddError(errors.Errorf("Error sending email for notifier '%s': %s",
+				notifierConfig.GetEmailConfig().GetNotifierId(), err))
+		}
+	}
+	if !errorList.Empty() {
+		s.logErrorAndUpsertReportStatus(errorList.ToError(), req)
+	}
+}
+
+func (s *scheduler) retryableSendReportResults(ctx context.Context, reportNotifier notifiers.ReportNotifier, mailingList []string,
+	zippedCSVData *bytes.Buffer, messageText string) error {
+	return retry.WithRetry(func() error {
+		return reportNotifier.ReportNotify(ctx, zippedCSVData, mailingList, messageText)
+	},
+		retry.OnlyRetryableErrors(),
+		retry.Tries(3),
+		retry.BetweenAttempts(func(previousAttempt int) {
+			wait := time.Duration(previousAttempt * previousAttempt * 100)
+			time.Sleep(wait * time.Millisecond)
+		}),
+	)
+}
+
+func (s *scheduler) notifySchedulerForMoreReports() {
+	s.sendNewReports <- true
+}
+
+func (s *scheduler) getReportData(ctx context.Context, rc *storage.ReportConfiguration, dataStartTime *types.Timestamp) ([]common.Result, error) {
```

Comment:
It's hard to follow the file since it mixes scheduling and report generation. You don't have to completely refactor for this PR (unless you have time), but I really recommend doing so soon

Reply:
Yeah, that is right. I split the report generation part and scheduling part into two files.